test/test_torch.py

```diff
@@ -2906,6 +2906,413 @@ def test_use_deterministic_algorithms_beta_warning(self, device):
             with self.assertWarnsOnceRegex(UserWarning, "torch.use_deterministic_algorithms is in beta"):
                 torch.use_deterministic_algorithms(True)
 
+    # This test checks that each nondeterministic function throws an error as
+    # described in the documentation for torch.use_deterministic_algorithms
+    @onlyOnCPUAndCUDA
+    def test_use_deterministic_algorithms_errors(self, device):
+
+        # Add new test cases to this list any time a nondeterministic error is
+        # added to an operation. Each entry has the following fields:
+        #
+        # name                  Name of test case
+        #
+        # operation             Operation that should throw nondeterministic error
+        #
+        # constructor           Function to construct args and kwargs
+        #
+        # error_device          Device that should throw the error.
+        #                       Set to None if all devices should throw an error.
+        #
+        # caller_name_forward   Name of function shown in error for forward pass.
+        #                       Set to None if the forward should not throw an error.
+        #
+        # caller_name_backward  Name of function shown in error for backward pass.
+        #                       Set to None if the backward should not throw an error.
+        nondeterministic_operations = [
```

Comment:
@mruberry, I realized I could simplify these entries a little bit and clean up the long lines by adding a constructor function that defines the args and kwargs to give to the forward function. This has a bit of a weird consequence that the nn modules need to specify their `.__call__` method for the `operation` parameter. Maybe this is bad? I'm not sure. We could potentially just use the `nn.functional` version of all the nn modules to make it look a little nicer.

Reply:
This looks OK to me. Whether we want to stick with this approach or just hand-write the "subtests" by creating the inputs, calling the op, and then verifying the error is thrown, is just a question of # LoC and readability/maintainability. For maintainability, consider that this will need to be documented in the pattern for adding a deterministic operation, and if people have questions about how to extend the test they'll come to you, so whatever minimizes those questions is probably the best thing to do.

Reply:
Good points. Considering the challenges of making the structured approach flexible enough to fit every case and simple enough to understand easily, handwritten tests seem like a better way to go